### PR TITLE
Add zfs

### PR DESCRIPTION
**Pull request by Mno-hime**
Based on work by @Aneurin (also used on ZoL).

Removed `zpool` code as we have our own.

### CODE REVIEW

File: zfs [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+# HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+# WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+
+if [[ -w /dev/zfs ]]; then
+    __ZFS_CMD="zfs"
+else
+    __ZFS_CMD="sudo zfs"
```
Comment (pyhalov) [inline zfs]: Why do we need sudo here?

**Comment by Aneurin** [comment]
Any chance you could add a link to https://github.com/Aneurin/zfs-bash in the vicinity of the copyright notice? I'm hopeful that might increase the chance of people from the future giving me a ping if they make any improvements.

Also, what problems did you have using the default awk and sed rather than gawk and sed? I'm wondering if I should make an update to check if they're available first and use them preferentially. I'm surprised though because I used to use this on OpenIndiana without any problems. I suppose it's possible I was adding `/usr/gnu/bin` to the beginning of my `$PATH` and just don't remember it.

Oh, FWIW the check for whether `/dev/zfs` is writable and switch to sudo is just a best-effort fallback rather than giving up entirely. IIRC it should always be writable on Illumos where the access permissions aren't handled by permissions on the device node, so it only really affects ZoL where there's a fairly high chance that it's only writable by root but also that there's a sudo setup allowing read-only commands.

**Comment by Mno-hime** [comment]
Thanks for reaching out @Aneurin! See the change in https://github.com/OpenIndiana/openindiana-completions/pull/10/commits/b67fffc91d8025600529510e5bc422f169ebe01c for the upstream URL note.

Wrt `awk` & `sed`: both illumos implementations lack features `gawk`/`nawk` and GNU `sed` have so we have to use them in the https://github.com/OpenIndiana/oi-userland/tree/oi/hipster/components/shell/bash-completion userland component (and other bash completions as well), and as I intend to ship https://github.com/OpenIndiana/openindiana-completions as part of bash-completion component I'd rather stick with `gawk` and `gsed` where possible even if not strictly necessary.

Thanks for the comment on `sudo` and `/dev/zfs`, I removed that code.

**Comment by pyhalov** [comment]
Re gsed and gawk - can we fix completions to work with nawk and our sed?

**Comment by pyhalov** [comment]
system/extended-system-utilities are installed in zone / minimal install by default, so we can avoid dependency on gawk

**Comment by Mno-hime** [comment]
I think it can be done for this project, but I don't think illumos `sed` can be used in https://github.com/OpenIndiana/oi-userland/tree/oi/hipster/components/shell/bash-completion/patches (remember I suggested to integrate this project to bash-completion component). And we have to use the same utilities both here and in the bash-completion component to be able to get rid of the GNU utilities. Or am I missing something? Maybe there's a better alternative to shipping these bits I fail to see?

**Comment by Mno-hime** [comment]
Or course, bash-completion component and this one can be changed to use `nawk` instead of `gawk`, but I am not very hopeful we can get rid of GNU sed nor GNU grep.

**Comment by Mno-hime** [comment]
I think I was negatively preoccupied with illumos `awk` and `sed`. I tested that `awk` works here, and could not make `sed` fail. Reverted to stock illumos utilities. Thanks.

**Comment by alarcher** [comment]
@Mno-hime Let me know if I can merge this or if you expect to modify it further.

**Comment by Mno-hime** [comment]
Unless further feedback I am done here.

**Pull request merged**